vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,414 @@
+	metricEncrypt = []string{"ocikms", "encrypt"}
+	metricDecrypt = []string{"ocikms", "decrypt"}
+
+	metricInitFailed    = []string{"ocikms", "initFailed"}
+	metricEncryptFailed = []string{"ocikms", "encryptFailed"}
+	metricDecryptFailed = []string{"ocikms", "decryptFailed"}
+)
+
+// OCIKMSMechanism is the method used to encrypt/decrypt in auto unseal process
+type OCIKMSMechanism uint32
+
+type OCIKMSSeal struct {
+	authTypeAPIKey bool   // true for user principal, false for instance principal, default is false
+	keyID          string // OCI KMS keyID
+
+	cryptoEndpoint     string // OCI KMS crypto endpoint
+	managementEndpoint string // OCI KMS management endpoint
+
+	cryptoClient     *keymanagement.KmsCryptoClient     // OCI KMS crypto client
+	managementClient *keymanagement.KmsManagementClient // OCI KMS management client
+
+	currentKeyID *atomic.Value // Current key version which is used for encryption/decryption
+
+	logger log.Logger
+}
+
+var _ seal.Access = (*OCIKMSSeal)(nil)
+
+// NewSeal creates a new OCIKMSSeal seal with the provided logger
+func NewSeal(logger log.Logger) *OCIKMSSeal {
+	k := &OCIKMSSeal{
+		logger:       logger,
+		currentKeyID: new(atomic.Value),
+	}
+	k.currentKeyID.Store("")
+	return k
+}
+
+func (k *OCIKMSSeal) SetConfig(config map[string]string) (map[string]string, error) {
+	defer metrics.MeasureSince(metricInit, time.Now())
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	// Check and set KeyID
+	switch {
+	case os.Getenv(EnvOCIKMSSealKeyID) != "":
+		k.keyID = os.Getenv(EnvOCIKMSSealKeyID)
+	case config[KMSConfigKeyID] != "":
+		k.keyID = config[KMSConfigKeyID]
+	default:
+		metrics.IncrCounter(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigKeyID)
+	}
+	k.logger.Info("OCI KMS configuration", KMSConfigKeyID, k.keyID)
+
+	// Check and set cryptoEndpoint
+	switch {
+	case os.Getenv(EnvOCIKMSCryptoEndpoint) != "":
+		k.cryptoEndpoint = os.Getenv(EnvOCIKMSCryptoEndpoint)
+	case config[KMSConfigCryptoEndpoint] != "":
+		k.cryptoEndpoint = config[KMSConfigCryptoEndpoint]
+	default:
+		metrics.IncrCounter(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigCryptoEndpoint)
+	}
+	k.logger.Info("OCI KMS configuration", KMSConfigCryptoEndpoint, k.cryptoEndpoint)
+
+	// Check and set managementEndpoint
+	switch {
+	case os.Getenv(EnvOCIKMSManagementEndpoint) != "":
+		k.managementEndpoint = os.Getenv(EnvOCIKMSManagementEndpoint)
+	case config[KMSConfigManagementEndpoint] != "":
+		k.managementEndpoint = config[KMSConfigManagementEndpoint]
+	default:
+		metrics.IncrCounter(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigManagementEndpoint)
+	}
+	k.logger.Info("OCI KMS configuration", KMSConfigManagementEndpoint, k.managementEndpoint)
+
+	// Check and set authTypeAPIKey
+	var err error
+	k.authTypeAPIKey = false
```

Comment:
This config item doesn't seem to be necessary.  You should just use this behavior if the key is present, otherwise fall back to instance auth.

Reply:
API Key can be shared for other features on the same host too. I think having this can to give user the flexibility to use instance auth even when some other features use API Key? What do you think?

Reply:
(Sorry I think I forgot to hit Submit for the comments)
OCI SDK doesn't enforce the hierarchy for access. It only provides the API for client to choose whether they want to use user/instance principal. Could you tell me why the bools are hard to configure?

Reply:
Makes sense.